### ISSUE TITLE
Feature Request: rsample for Von Mises distribution

### ISSUE THREAD

**Issue by dario-loi** [pytorch/pytorch#147048]
## 🚀 Feature

The von Mises-Fisher distribution implemented in `torch.distribution` should get an `.rsample()` method.

## Motivation

Backpropagating through vMF is essential to train Hyperspherical VAEs, which have drastically better performance for directional data, for example in graph reconstruction (https://arxiv.org/abs/1804.00891).

Without `.rsample` one is limited to gaussian priors for VAE training.

## Additional context

An implementation that exposes an `.rsample()` method is available [in this repo](https://github.com/nicola-decao/s-vae-pytorch/blob/master/hyperspherical_vae/distributions/von_mises_fisher.py).

**Comment by giovannicozzolongo** [comment]
`torch.distributions.VonMises` is the scalar circular distribution, `event_shape == ()`. The body of this issue and the paper it cites are about the vector-valued von Mises-Fisher on the sphere, which is #13811. So the title and the body want different things.

On the implementation linked above, I would not copy it. For d other than 3 it differentiates straight through a Wood/Ulrich rejection sampler without the loss-weighted correction from Naesseth et al. (2017), so the pathwise gradient is biased. At d = 5, kappa = 2, for `d E[mu^T x] / d kappa`:

```
pathwise, linked implementation   0.127662 +/- 0.000079
analytic                          0.147389
score function                    0.147510 +/- 0.000271
```

That is about 13% low and 249 standard errors out, and the score-function estimate never touches the sampler gradient. TensorFlow Probability only claims full reparameterisation for its d = 3 inverse-CDF sampler.

@fritzo, in #13811 you recommended ProjectedNormal over vMF for exactly this reason. The open part is the point dario-loi raised there and never got an answer to: without a closed-form KL against a projected standard normal, ProjectedNormal does not cover the VAE use case that motivates both issues. Is ProjectedNormal still the preferred direction, or would a 3D-only vMF be worth having in core? At d = 3 the normalising constant is elementary and the CDF inverts in closed form, so there is no rejection step and `rsample()` is reparameterised by construction. I can put that together if it is wanted.

**Comment by dario-loi** [comment]
Consider the Power Spherical distribution as an object providing similar behavior (prior on $S^{d - 1}$) with well defined KL-divergence and an analytical inverse CDF, avoiding rejection sampling

ICML 2020 INNF Workshop Presentation: https://icml.cc/virtual/2020/7754
Paper: https://arxiv.org/html/2006.04437
